### PR TITLE
🧪  Layer squashing, a uniqueness `BONRefiner`, some early network vignette functions

### PR DESCRIPTION
**Pull request by gottacatchenall**
Main things in this PR:

- A new `BONRefiner` that picks the N-most unique locations from a seeded set of coordinates across a set of layers passed to it
- Optimization of environmental uniqueness as a vignette (in `./docs/vignettes/uniqueness.jl`)
- The `squash` function, for taking a weight matrix and vector and collapses many layers into one.
- Lazy-loading of the SimpleSDMLayers, NeutralLandscapes, and Zygote packages. 
    - SimpleSDMs: Integration with `missing` values in maps
    - Zygote: Implementation of AD gradient descent  
- Simulation of metaweb sampling functions for the forthcoming network vignette (in `./docs/vignettes/networks.jl`)

### CODE REVIEW

**Comment by gottacatchenall** [comment]
Should we merge this and implement batching in a new PR? @tpoisot

**Comment by tpoisot** [comment]
Yeah

File: misc/metaweb_generation.jl [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# ===============================================
```
Comment (tpoisot) [inline misc/metaweb_generation.jl:1]: shouldn't this be a vignette?

File: src/optimize.jl [hunk 2]
```diff
@@ -0,0 +1,82 @@
+"""
```
Comment (tpoisot) [inline src/optimize.jl:1]: docstring isn't updated

File: src/optimize.jl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    to reduce a loss function that describes the 
+    difference between the "true" metaweb and 
+    sampled one. 
+
+    In our context, the two targets are interaction classification
+    and network topology, so we want a loss function that combines
+    measures of these elements.
+"""
+
+struct Weights{F <: AbstractFloat}
```
Comment (tpoisot) [inline src/optimize.jl:30]: is it actually used?

File: src/optimize.jl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        α ./= sum(α)
+
+        losses[step] = loss(W, α)
+    end
+    return losses
+end
+
+# ...?
+
+using Statistics, StatsBase
```
Comment (tpoisot) [inline src/optimize.jl:56]: drop this from the file

File: src/optimize.jl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+function _squish(layers::Array{T, 3}, W::Matrix{T}) where {T <: AbstractFloat}
+    return convert(Array, slicemap(x -> x * W, layers; dims = (2, 3)))
+end
+
+function _squish(layers::Array{T, 3}, α::Vector{T}) where {T <: AbstractFloat}
+    return slicemap(x -> x * reshape(α, (length(α), 1)), layers; dims = (2, 3))[:, :, 1]
+end
+
+dims, nl, nt = (50, 50), 5, 3
```
Comment (tpoisot) [inline src/optimize.jl:70]: drop this from the file (move to tests?)

**Review by tpoisot — CHANGES_REQUESTED** [review]

**Comment by gottacatchenall** [comment]
I'm planning to add a new `BONRefiner` called `Uniqueness` that selects the `n`-sites with the least mean covariance with other sites, e.g. if total covar per site is calculated as $C_i  = \sum_j \text{Cov}(i,j)$, this `Refiner` takes `N` points and returns the `m` points with the smallest values of $C\_i$.

This creates an issue in how the api would work, because we want to pass `Uniqueness` a set of environmental layers (An Array{T,3}, instead of a Matrix), e.g.

```
candidates = zeros(50,50) |> seed(AdaptiveSampling())
refine(candidates, Uniqueness(), layers)
```

Two solution I can see for this are

(1) dropping the type restriction on `uncertainty` as an matrix and doing an argument check inside `_generate`

or 

(2) store layers as a field in `Uniqueness` and ignore the `uncertainty` matrix passed to it by `refine`

@tpoisot

**Comment by gottacatchenall** [comment]
i went with (2)

**Comment by gottacatchenall** [comment]
seems ready for review @tpoisot

**Review by tpoisot — APPROVED** [review]

**Comment by tpoisot** [comment]
Do you want to tag when merged?

**Comment by gottacatchenall** [comment]
yeah sure

**Pull request merged**